### PR TITLE
Forward Flow/Lexical Cohesion

### PR DESCRIPTION
**Pull request by agshruti12**
_Pull Request Template_:
If you are merging in a feature or other major change, use this template to check your pull request!

# Basic Info
What's this pull request about? 
> Implementing 2 versions of forward flow (pairwise, cumulative).

# Feature Documentation
Did you document your feature? Make sure you do the following before you pull request!
- [ ] Copy the Template. Go to the Feature Wiki and Copy/Paste the Feature Template into a new page.
- [ ] Fill out the Template. Fill out the basic information for the feature in the template. Use the template to document your plan for implementation and major design decisions; if anything changes along the way, update the documentation as you go.
- [ ] At the top of my feature, I indicate whether the feature is *conversation level* or *chat level*.

# Code Basics
- [ ] My feature is a .py file.
- [ ] My feature uses snake case in the name. That means the name of the format is `my_feature`, NOT `myFeature` (camel case).
- [ ] My feature has the name, `NAME_features.py`, where NAME is the name of my feature.
- [ ] My feature is located in `feature_engine/features`.

# Testing
- [ ] I have thought about test cases for my features, with inputs and expected outputs.
- [ ] I have linked to a location (e.g., .py or .ipynb) where I can run my test cases and show they work (inputs match expected outputs).

The location of my tests are here:
> testing/fflow.ipynb

If you check all the boxes above, then you ready to merge!

### CODE REVIEW

File: feature_engine/features/fflow.py [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+
+        for i, pair in conv.groupby(conv.index // 2):
+            
+            # last "pair" has only one element, safeguard against this
+            if (len(pair) == 2):
+                
+                cos_sim_matrix = cosine_similarity([pair.iloc[0]['message_embedding'], pair.iloc[1]['message_embedding']])
+                lex_cohesion_adj_chats.append(cos_sim_matrix[np.triu_indices(len(cos_sim_matrix), k = 1)][0])
+                cached_cohesion += cos_sim_matrix[np.triu_indices(len(cos_sim_matrix), k = 1)][0]
+                lex_cohesion_cumulative.append(cached_cohesion/pair_index)
```
Comment (xehu) [inline feature_engine/features/fflow.py:29]: These are essentially calculating chat-level values and then summarizing them later (in lines 32-33), so let's move this to the chat level instead (as we discussed in the meeting)!

Let me know if any part of it gets confusing; happy to step in! :)

**Review by xehu — CHANGES_REQUESTED** [review]

File: feature_engine/features/lexicons/certainty.txt
Comment (xehu) [inline feature_engine/features/lexicons/certainty.txt:1]: Can we remove the certainty.txt file from this PR, as it's not related to forward flow?

Also, we need to add it to the .gitignore, as we don't want to commit the actual file to the main branch.

**Review by xehu — CHANGES_REQUESTED** [review]

File: feature_engine/features/lexicons/certainty.txt
Comment (xehu) [reply feature_engine/features/lexicons/certainty.txt:1]: Update: I added the Certainty .txt to the gitignore in the latest merge! Let's just make sure it's deleted from this file.

**Comment by agshruti12** [comment]
Just fixed forward flow, and added "mimicry bert" and "moving mimicry" to the chat level feature pipeline.

**Review by xehu — APPROVED** [review]

**Pull request merged**